packages/block-editor/src/hooks/use-bindings-attributes.js

```diff
@@ -32,69 +32,101 @@ const BLOCK_BINDINGS_ALLOWED_BLOCKS = {
 
-							if ( metaValue ) {
-								updatedAttributes[ attributeName ] = metaValue;
-							}
-						}
+			const boundAttributes = {};
+			if ( attributes?.metadata?.bindings ) {
+				Object.entries( attributes?.metadata?.bindings ).forEach(
+					( [ attribute, binding ] ) => {
+						boundAttributes[ attribute ] = getBlockBindingsSource(
+							binding.source
+						).useSource( props, binding.args );
 					}
 				);
 			}
 
+			const attributesWithBindings = useMemo( () => {
+				return {
+					...attributes,
+					...Object.fromEntries(
+						BLOCK_BINDINGS_ALLOWED_BLOCKS[ name ].map(
+							( attributeName ) => {
+								// Check bindings.
+								if ( boundAttributes[ attributeName ] ) {
+									const {
+										placeholder,
+										useValue: [ sourceValue = null ] = [],
+									} = boundAttributes[ attributeName ];
+
+									const blockTypeAttribute =
+										blockType.attributes[ attributeName ];
+
+									if ( placeholder && ! sourceValue ) {
+										// If the attribute is `src` or `href`, a placeholder can't be used because it is not a valid url.
+										// Adding this workaround until attributes and metadata fields types are improved and include `url`.
+
+										const htmlAttribute =
+											blockTypeAttribute.attribute;
+										if (
+											htmlAttribute === 'src' ||
+											htmlAttribute === 'href'
+										) {
+											return [ attributeName, null ];
+										}
+										return [ attributeName, placeholder ];
+									}
+
+									if ( sourceValue ) {
+										// TODO: If it is rich-text, I think we can't edit it this way.
+										return [ attributeName, sourceValue ];
+									}
+								}
+								return [
+									attributeName,
+									attributes[ attributeName ],
+								];
+							}
+						)
+					),
+				};
+			}, [ attributes, blockType.attributes, boundAttributes, name ] );
+
+			const updatedSetAttributes = useCallback(
+				( nextAttributes ) => {
+					Object.entries( nextAttributes ?? {} )
+						.filter(
+							( [ attribute ] ) => attribute in boundAttributes
+						)
+						.forEach( ( [ attribute, value ] ) => {
+							const {
+								useValue: [ , setSourceValue = null ] = [],
+							} = boundAttributes[ attribute ];
+							if ( setSourceValue ) {
+								setSourceValue( value );
+							}
+						} );
+					setAttributes( nextAttributes );
+				},
+				[ setAttributes, attributesWithBindings, boundAttributes ]
+			);
+
 			const registry = useRegistry();
 
 			return (
-				<>
-					<BlockEdit
-						key="edit"
-						attributes={ updatedAttributes }
-						setAttributes={ ( newAttributes, blockId ) =>
-							registry.batch( () =>
-								updateBlockAttributes( blockId, newAttributes )
-							)
-						}
-						{ ...props }
-					/>
-				</>
+				<BlockEdit
+					attributes={ attributesWithBindings }
+					setAttributes={ ( newAttributes ) => {
+						registry.batch( () =>
+							updatedSetAttributes( newAttributes )
+						);
+					} }
```

Comment:
If I'm right, this scenario will cover the case when the attributes are edited from the BlockEdit function representation. This means that if, for instance, you change the attribute value by dispatching an action, the updatedSetAttribute will never be called.